src/rust/engine/src/downloads.rs

```diff
@@ -90,6 +96,108 @@ impl StreamingDownload for NetDownload {
   }
 }
 
+struct S3Download {
+  stream: futures_core::stream::BoxStream<'static, Result<Bytes, Error>>,
+}
+
+impl S3Download {
+  fn get_bucket_and_key(url: &Url) -> (String, String) {
+    let mut path = url.path().to_string();
+    path.remove(0);
+    (url.host_str().unwrap().to_string(), path)
+  }
+
+  async fn get_aws_credentials() -> Result<Credentials, StreamingError> {
+    let config = aws_config::load_from_env().await;
+    let credential_provider = config
+      .credentials_provider()
+      .expect("a credentials provider is loaded");
+    let maybe_credentials = credential_provider.provide_credentials().await;
+    let credentials = maybe_credentials
+      .map_err(|err| StreamingError::Permanent(format!("Error loading AWS credentials: {err}")))?;
+    Ok(credentials)
+  }
+
+  async fn get_headers(bucket: &String, key: &String) -> Result<header::HeaderMap, StreamingError> {
+    let content_type = "binary/octet-stream";
+    let date = Utc::now().to_rfc2822();
+    let resource = format!("/{bucket}/{key}");
+    let unsigned_signature = format!("GET\n\n{content_type}\n{date}\n{resource}");
+
+    let credentials = S3Download::get_aws_credentials().await?;
+    let mut mac = Hmac::new(Sha1::new(), credentials.secret_access_key().as_bytes());
+    mac.input(unsigned_signature.as_bytes());
+    let result = mac.result();
+    let signature_hash = base64::encode(result.code());
+
+    let mut headers = header::HeaderMap::new();
+    headers.insert(
+      header::AUTHORIZATION,
+      format!("AWS {}:{}", credentials.access_key_id(), signature_hash)
+        .parse()
+        .unwrap(),
+    );
+    headers.insert(header::DATE, date.parse().unwrap());
+    headers.insert(header::CONTENT_TYPE, content_type.parse().unwrap());
+    Ok(headers)
+  }
+
+  async fn start(
+    core: &Arc<Core>,
+    url: Url,
+    file_name: String,
+  ) -> Result<S3Download, StreamingError> {
+    let (bucket, key) = S3Download::get_bucket_and_key(&url);
+    let url = env::var("TESTING_PANTS_S3_FMT_URL") // NB: We set this for testing
+      .unwrap_or_else(|_| "https://{}.s3.amazonaws.com/{}".to_string())
+      .replacen("{}", &bucket, 1)
+      .replacen("{}", &key, 1);
+
+    let response = core
+      .http_client
+      .get(url.clone())
+      .headers(S3Download::get_headers(&bucket, &key).await?)
```

Comment:
Everything after this is shared with the normal download code. I'll probably wanna helper this